lua_modules/ftp/ftpserver.lua

```diff
@@ -0,0 +1,182 @@
+      return c:send("250 OK. Current directory is "..cwd.."\r\n")
+    end
+    if a1 == "CWD" then
+      if a2 == "." then
+        return c:send("257 \""..cwd.."\" is your current directory\r\n")
+      end
+      if a2 == "/" then
+        cwd = a2
+        return c:send("250 OK. Current directory is "..cwd.."\r\n")
+      end
+      return c:send("550 Failed to change directory.\r\n")
+    end
+    if a1 == "PWD" then
+      return c:send("257 \""..cwd.."\" is your current directory\r\n")
+    end
+    if a1 == "TYPE" then
+      if a2 == "A" then
+        t = 0
+        return c:send("200 TYPE is now ASII\r\n")
+      end
+      if a2 == "I" then
+        t = 1
+        return c:send("200 TYPE is now 8-bit binary\r\n")
+      end
+      return c:send("504 Unknown TYPE")
+    end
+    if a1 == "MODE" then
+      if a2 ~= "S" then
+        return c:send("504 Only S(tream) is suported\r\n")
+      end
+      return c:send("200 S OK\r\n")
+    end
+    if a1 == "PASV" then
+      local _,ip = c:getaddr()
+      local _,_,i1,i2,i3,i4 = string.find(ip,"(%d+).(%d+).(%d+).(%d+)")
+      return c:send("227 Entering Passive Mode ("..i1..","..i2..","..i3..","..i4..",4,0).\r\n")
+    end
+    if a1 == "LIST" or a1 == "NLST" then
+      c:send("150 Accepted data connection\r\n")
+      data_fnc = function(sd)
+        local l = file.list();
+        for k,v in pairs(l) do
+          sd:send( a1 == "NLST" and k.."\r\n" or "-rwxrwxrwx 1 esp esp "..v.." Jan  1  2018 "..k.."\r\n")
+        end
+        sd:close()
+        data_fnc = nil
+        c:send("226 Transfer complete.\r\n")
+      end
+      if data_sock then
+        node.task.post(function() data_fnc(data_sock);data_sock=nil end)
+      end
+      return
+    end
+    if a1 == "RETR" then
+      local f = file.open(a2:gsub("%/",""),"r")
+      if f == nil then
+        return c:send("550 File "..a2.." not found\r\n")
+      end
+      c:send("150 Accepted data connection\r\n")
+      data_fnc = function(sd)
+        sd:on("sent", function(cd)
+          b=f:read(1024)
+          if b then
+            cd:send(b)
+            b=nil
+          else
+            cd:close()
+            f:close()
+            data_fnc = nil
+            c:send("226 Transfer complete.\r\n")
+          end
+        end)
+        local b=f:read(1024)
+        sd:send(b)
+        b=nil
+      end
+      if data_sock then
+        node.task.post(function() data_fnc(data_sock);data_sock=nil end)
+      end
+      return
+    end
+    if a1 == "STOR" then
+      local f = file.open(a2:gsub("%/",""),"w")
+      if f == nil then
+        return c:send("451 Can't open/create "..a2.."\r\n")
+      end
+      c:send("150 Accepted data connection\r\n")
+      data_fnc = function(sd)
+        sd:on("receive", function(cd, dd)
+          f:write(dd)
+        end)
+        sd:on("disconnection", function(c)
+          f:close()
+          data_fnc = nil
+        end)
+        c:send("226 Transfer complete.\r\n")
```

Comment:
The ``c:send("226 Transfer complete.\r\n")`` has to be executed after disconnect only. Else it stops sending the file after a couple of blocks which results in cut off files.

The ``on disconnection`` should look loke this: (also note the rename of the function param from ``c`` to ``cd``)

```
        sd:on("disconnection", function(cd)
          f:close()
          data_fnc = nil
          c:send("226 Transfer complete.\r\n")
        end)
```